src/java.base/share/classes/sun/security/provider/HSS.java

```diff
@@ -0,0 +1,744 @@
+            n = lmotsParams.n;
+            p = lmotsParams.p;
+            if (inLen != (4 + n * (p + 1)))
+                throw new InvalidParameterException("OTS signature has incorrect length");
+            C = Arrays.copyOfRange(sigArray, 4, 4 + n);
+            int pStart = 4 + n;
+            y = new byte[p][n];
+            for (int i = 0; i < p; i++) {
+                y[i] = Arrays.copyOfRange(sigArray, pStart, pStart + n);
+                pStart += n;
+            }
+        }
+
+        public void getC(byte[] arr, int pos) {
+            System.arraycopy(C, 0, arr, pos, n);
+        }
+
+        public void getY(int i, byte[] arr, int pos) {
+            System.arraycopy(y[i], 0, arr, pos, n);
+        }
+    }
+
+    static class LMParams {
+        final int type;
+        final int m; // the number of bytes used from the hash output
+        final int hashAlg_m = 32; // output length of the hash function used for the LMS tree
+        final int h; // height of the LMS tree
+        final int twoPowh;
+        final String hashAlgStr;
+
+        LMParams(int type) {
+            this.type = type;
+            switch (type) {
+                case LMSUtils.LMS_SHA256_M32_H5:
+                    m = 32;
+                    h = 5;
+                    hashAlgStr = "SHA-256";
+                    break;
+                case LMSUtils.LMS_SHA256_M32_H10:
+                    m = 32;
+                    h = 10;
+                    hashAlgStr = "SHA-256";
+                    break;
+                case LMSUtils.LMS_SHA256_M32_H15:
+                    m = 32;
+                    h = 15;
+                    hashAlgStr = "SHA-256";
+                    break;
+                case LMSUtils.LMS_SHA256_M32_H20:
+                    m = 32;
+                    h = 20;
+                    hashAlgStr = "SHA-256";
+                    break;
+                case LMSUtils.LMS_SHA256_M32_H25:
+                    m = 32;
+                    h = 25;
+                    hashAlgStr = "SHA-256";
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unsupported or bad LMS type");
+            }
+
+            twoPowh = 1 << h;
+        }
+
+    }
+
+    static class LMSignature {
+        final int sigLmType;
+        final int sigOtsType;
+        final private byte[] qArr;
+        final int q; // serial number of the LMS key being used for this signature
+        final LMOTSignature lmotSig;
+        final int n; // output length of the hash function used in the one time signature
+        final int p; // number of hash chains in the signature
+        final int m; // output length of the hash fubction used in the Merkle tree
+        final int h; // height of the Merkle tree
+        final byte[][] path;
+        final byte[] sigArr;
+
+        public LMSignature(byte[] sigArray, int offset, boolean checkExactLen) throws SignatureException {
```

Comment:
The methods of `LMSignature` can be package-private.

Reply:
Changed.